File: EstructurasDeDatosII/PythonNayelli/Arboles/BST.py

class Node:

    def __init__(self,data=0):
        self.left = None
        self.right = None
        self.data = data

    def __str__(self):
        return str(self.data)


class BST:
    
    def __init__(self):
        self.root = None

    def insert(self, data):
        n = Node(data)
        aux = self.root
        if self.root == None:
            self.root = n
            return
        while(True):
            if data <= aux.data:  # Izquierda
                if aux.left == None:
                    aux.left = n
                    return
                else:
                    aux = aux.left
            else:                 # Derecha
                if aux.right == None:
                    aux.right = n
                    return
                else:
                    aux = aux.right

    def nodoChiquito(self,root):
        while root.right != None:
            root = root.right
        return root

    def delete(self, data, esteSi = None):
        if self.root == None:
            return
        ant = None
        aux = self.root

        while esteSi != None and aux != esteSi and aux.data == data:
            ant = aux
            if data <= aux.data:
                aux = aux.left
            else:
                aux = aux.right

        while aux != None and aux.data != data:
            ant = aux
            if data <= aux.data:
                aux = aux.left
            else:
                aux = aux.right
            
            while esteSi != None and aux != esteSi and aux.data == data:
                ant = aux
                if data <= aux.data:
                    aux = aux.left
                else:
                    aux = aux.right

        # NO ESTA - REGRESO
        if aux == None:
            return

        # CASO 1 : HOJA
        if aux.left == None and aux.right == None and aux.data == data:
            # Si mi root es mi hoja, un solo elemento 
            if aux == self.root:
                self.root = None
            else:  # Es algun otro nodo, checo con el anterior
                if ant.left == aux:
                    ant.left = None
                else:
                    ant.right = None
            del aux

        # Caso 3 : DOS HIJOS
        elif aux.right != None and aux.left != None:
            # Saco el nodo mas pequeño que es mas grande que el que tengo
            nodoCambio = self.nodoChiquito(aux.left)
            # Y lo elimino, pero guardo el valor antes
            valNodo = nodoCambio.data
            # Recursivamente uso la misma funcion delete 
            self.delete(valNodo, nodoCambio)
            # Y cambio el valor por elvalor del nodo eliminado
            aux.data = valNodo
            
        # Caso 2 : 1 hijo
        else:
            # Es parecido al caso de arriba :p
            hijo = 0
            if aux.left == None: # hay hijo derecho
                hijo = aux.right
            else:    # Hay hijo izquierdo
                hijo = aux.left

            # Si mi valor es root
            if aux == self.root:
                self.root = hijo
            else:
                if aux == ant.left: # Si originalmente estaba a la izquierda
                    ant.left = hijo
                else:  # Era el derecho
                    ant.right = hijo
            del aux

        return

    def DFS(self):
        st = "["
        s = list()
        s.append(self.root)
        while s != []:
            n = s.pop()
            if n != self.root: st+=", "
            st+=str(n.data)
            if n.right != None: s.append(n.right)
            if n.left != None: s.append(n.left)
        st+="]"
        return st

    def BFS(self):
        st = "["
        q = list()
        q.append(self.root)
        while q != []:
            n = q[0]
            q = q[1:]
            if n != self.root: st+=", "
            st+=str(n.data)
            if n.left != None: q.append(n.left)
            if n.right != None: q.append(n.right)
        st+="]"
        return st

    def contains(self, data):
        aux = self.root
        while(aux != None):
            if aux.data == data:
                return True
            elif data <= aux.data:  # Izq
                aux = aux.left
            else:
                aux = aux.right
        return False

    def preorder(self,root):
        if root != None:
            v = []
            v.append(root.data)
            x = self.preorder(root.left)
            if x != None:
                v+=x
            y = self.preorder(root.right)
            if y != None:
                v+=y

            return v
        return
        
    def inorder(self,root):
        if root != None:
            v = []
            x = self.inorder(root.left)
            if x != None:
                v+=x
            v.append(root.data)
            y = self.inorder(root.right)
            if y != None:
                v+=y

            return v
        return

    def postorder(self,root):
        if root != None:
            v = []
            x = self.postorder(root.left)
            if x != None:
                v+=x
            y = self.postorder(root.right)
            if y != None:
                v+=y
            v.append(root.data)

            return v
        return
    
    def printNode(self,root,space=""):
        if root == None: return ""
        s = space+str(root.data)+"\n"
        space+=" "
        s+=self.printNode(root.left,space)
        s+=self.printNode(root.right,space)
        return s

    def mostrarArbol(self,root,cont=0):
        if root == None:
            return
        self.mostrarArbol(root.right,cont+1)
        s=""
        for i in range(cont):
            s+=" -"
        print(s+str(root.data))
        self.mostrarArbol(root.left,cont+1)

    def __str__(self):
        s = "Preorder - "+str(self.preorder(self.root)) +"\n"
        s+= "Inorder - "+str(self.inorder(self.root)) +"\n"
        s+= "Postorder - "+str(self.postorder(self.root)) +"\n"
        s+= "DFS - "+str(self.DFS()) +"\n"
        s+= "BFS - "+str(self.BFS()) +"\n"
        self.mostrarArbol(self.root)
        ##return self.printNode(self.root)
        return s

File: EstructurasDeDatosII/C++Nayelli/Arboles/BST.cpp
#include<bits/stdc++.h>
using namespace std;

template <class T>
class BST{
public:

    template <class H>
    class Node{
    public:
        Node<H> *left = nullptr, *right = nullptr;
        H data = (H)NULL;

        Node<H>(){}
        
        Node<H>(H data){
            this->data = data;
        }
            
        ~Node<H>(){
            cout<<"~Delete: "<<data<<'\n';
            return;
        }

    };
    
    Node<T>* root = nullptr;

    BST(){}

    ~BST(){
        cout<<"~Delete BST"<<'\n';
        toposort(this->root);
    }

    void insert(T data){
        Node<T>* n = new Node<T>(data);
        Node<T>* aux = this->root;
        if(aux == nullptr){
            this->root = n;
            return;
        }
        while(true){
            if(data <= aux->data){
                if(aux->left == nullptr){
                    aux->left = n;
                    return;
                }else{
                    aux = aux->left;
                }
            }else{
                if(aux->right == nullptr){
                    aux->right = n;
                    return;
                }else{
                    aux = aux->right;
                }
            }
        }
    }

    bool contains(T data){
        Node<T>* n = this->root;
        while(n != nullptr){
            if(n->data == data){
                return true;
            }else if(data <= n->data){
                n = n->left;
            }else{
                n = n->right;
            }
        }
        return false;
    }

    Node<T>* nodoChiquito(Node<T>* root){
        while(root->right != nullptr){
            root = root->right;
        }
        return root;
    }

    void remove(T data, Node<T>* esteSi = nullptr){
        // ROOT = NULL
        if(this->root == nullptr){
            return;
        }
        Node<T>* ant = nullptr;
        Node<T>* aux = this->root;
        while(esteSi != nullptr && aux != esteSi && aux->data == data){
            ant = aux;
            if(data <= aux->data){
                aux = aux->left;
            }else{
                aux = aux->right;
            }
        }
        while(aux != nullptr && aux->data != data){
            ant = aux;
            if(data <= aux->data){
                aux = aux->left;
            }else{
                aux = aux->right;
            }
            while(esteSi != nullptr && aux != esteSi && aux->data == data){
                ant = aux;
                if(data <= aux->data){
                    aux = aux->left;
                }else{
                    aux = aux->right;
                }
            }
        }
        // NO ESTA - REGRESO
        if(aux == nullptr){
            return;
        }

        // CASO 1 : HOJA
        if(aux->left == nullptr && aux->right == nullptr && aux->data == data){
            // Si es root, reinicio todo el arbol
            if(aux == this->root){
                this->root = nullptr;
            }else{ // El anterior lo contiene
                if(ant->left == aux){
                    ant->left = nullptr;
                }else{
                    ant->right = nullptr;
                }
            }
            delete aux;
        }

        // CASO 3 : DOS HIJOS
        else if(aux->right != nullptr && aux->left != nullptr){
            // Sacar nodo mas pequeño mas grande del que tengo
            Node<T>* nodoCambio = nodoChiquito(aux->left);
            // Y lo elimino, pero guardo el valor antes
            T valNodo = nodoCambio->data;
            // Recursivamente uso la misma funcion delete 
            remove(valNodo, nodoCambio);
            // Y cambio el valor por elvalor del nodo eliminado
            aux->data = valNodo;
            // No elimino el nodo, porque recursivamente elimino
            // el nodo que cotenia el valor anterior
        }

        // CASO 2 : 1 HIJO
        else{

            // Sacar el unico hijo
            Node<T>* hijo;
            if(aux->left != nullptr){
                hijo = aux->left;
            }else{
                hijo = aux->right;
            }

            // Si es mi root
            if(aux == this->root){
                this->root = hijo;
            }else{
                if(aux == ant->left){
                    ant->left = hijo;
                }else{
                    ant->right = hijo;
                }
            }

            delete aux;
        }
        
        return;
    }
    
    void mostrarArbol(Node<T>* root , int cont=0){
        if(root == nullptr){
            return;
        }
        mostrarArbol(root->right, cont+1);
        for(int i=0;i<cont;i++)cout<<"  ";
        cout<<root->data<<'\n';
        mostrarArbol(root->left, cont+1);

    }
    
    void toposort(Node<T>* root){
        if(root == nullptr) return;
        toposort(root->left);
        toposort(root->right);
        delete root;
        return;
    }

    void DFS(){
        stack<Node<T>*> st; 
        st.push(this->root);
        cout<<"[";
        while(!st.empty()){
            Node<T>* n = st.top();
            st.pop();
            if(n != this->root){
                cout<<", ";
            }
            cout<<n->data;
            if(n->right != nullptr) st.push(n->right);
            if(n->left != nullptr) st.push(n->left);
        }
        cout<<"]\n";
        return;
    }

    void BFS(){
        queue<Node<T>*> st; 
        st.push(this->root);
        cout<<"[";
        while(!st.empty()){
            Node<T>* n = st.front();
            st.pop();
            if(n != this->root){
                cout<<", ";
            }
            cout<<n->data;
            if(n->left != nullptr) st.push(n->left);
            if(n->right != nullptr) st.push(n->right);
        }
        cout<<"]\n";
        return;
    }

    void preorder(Node<T>* root, bool ya = false){
        if(root == nullptr) return;
        if(root== this->root)cout<<"[";
        if(ya)cout<<", ";
        cout<<root->data;
        preorder(root->left,1);
        preorder(root->right,1);
        if(root== this->root)cout<<"]\n";
        return;
    }
    
    bool inorder(Node<T>* root, bool ya = false){
        if(root == nullptr) return ya;
        if(root== this->root)cout<<"[";
        ya = max(ya,inorder(root->left,ya));
        if(ya)cout<<", ";
        cout<<root->data;
        ya = true;
        inorder(root->right,1);
        if(root== this->root)cout<<"]\n";
        return 1;
    }

    bool postorder(Node<T>* root, bool ya = false){
        if(root == nullptr) return 0;
        if(root== this->root)cout<<"[";
        ya = max(postorder(root->left,ya),ya);
        ya = max(postorder(root->right,ya),ya);
        if(ya)cout<<", "; 
        cout<<root->data;
        ya = true;
        if(root== this->root)cout<<"]\n";
        return ya;
    }

    void print(){
        DFS();
        BFS();
        preorder(this->root);
        inorder(this->root);
        postorder(this->root);
        mostrarArbol(this->root);
    }

};

int main(){
    
    BST<int> bst;
    bst.insert(5);
    bst.insert(4);
    bst.insert(6);
    bst.insert(3);
    bst.insert(9);
    bst.insert(1);
    bst.insert(8);
    bst.insert(3);
    bst.insert(3);
    bst.insert(3);
    bst.insert(3);
    bst.insert(3);
    bst.insert(7);
    bst.insert(0);
    bst.insert(2);
    bst.mostrarArbol(bst.root);

    bst.remove(4);
    bst.mostrarArbol(bst.root);
    bst.remove(2);
    bst.mostrarArbol(bst.root);
    bst.remove(3);
    bst.remove(3);
    bst.mostrarArbol(bst.root);
    bst.print();

    BST<char> b;
    b.insert('A');
    b.insert('B');
    b.insert('A');
    b.insert('B');
    b.remove('A');
    b.print();
}
